### ISSUE TITLE
Cancel async Cmd.run

### ISSUE THREAD

**Issue by ercpereda** [redux-loop/redux-loop#134]
Is there a way of avoiding the success action being called? I would like a way of canceling the effect of async call under certain conditions.

**Comment by bdwain** [comment]
Not directly. 

You could check this condition in the reducer where you handle this action and do nothing if the condition is true. 

You could also check the condition right before resolving the main promise and add the result to the resolution value, which gets passed to the action creator. Then you could check the value in the reducer from there.

**Comment by ercpereda** [comment]
Thanks, @bdwain. I thought similar solutions.
I think that the api could add a way to handle this scenario, that would be very helpful.
What do you think?

**Comment by ashishsc** [comment]
I wonder if its necessary. In the past, I have supported cancelling by just handling it in the reducer like @bdwain suggests. If you really want, you could cancel the promise if using a lib like Bluebird, which should dispatch the fail effect I think.

**Comment by ercpereda** [comment]
I really don't need the feature. I just wondering if would be useful have it in the api. `redux-loop` is my favorite tool to handle async operations in `redux`, I just want to help to improve the api.

**Comment by bdwain** [comment]
So the only api change i could think to accomplish it would be to add some optional function that checks if the promise should cancel. but that could just be done from the action creator. like if the action creator was fooSuccess, then do

```js
Cmd.run(foo, {
  successActionCreator: maybeFooSuccess
})

function maybeFooSuccess(res){
  return res.cancel ? null : fooSuccess(res)
}
```

i have to check if the action creator is allowed to return null. if not we could change it to do that and just ignore the null ones.

**Comment by ercpereda** [comment]
@bdwain the problem with that solution is that you still have to cancel your promise by you own, and the api is not helping with that. A way to go could keep track of the `Cmd.run` that are not being resolved yet and add a new `Cmd` to cancel it.

**Comment by bdwain** [comment]
so you want to be able to cancel any Cmd without the Cmd having to do anything to be cancellable? 

how would Cmd.cancel (or whatver) target a specific command?

**Comment by ercpereda** [comment]
Maybe adding a key option to the command
```javascript
Cmd.run(foo, {
  successActionCreator,
  failActionCreator,
  key: 'foo'
}

Cmd.cancel('foo')
```

**Comment by bdwain** [comment]
that would be a huge internal change, and still requires you to do something to a Cmd to make it cancellable. if you need to do something anyway, i'd rather wrap the successActionCreator than add a key to each command. It seems to accomplish the same goal without increasing the API's complexity. Is there something that can't be done using that method that they key method allows you to do?

**Comment by ercpereda** [comment]
I know that is not an easy implementation. It will be very useful in many scenarios. For example, an async autocomplete, even if you use something like debounce, you maybe should want to cancel all the old request before starting the new ones. There is no way to cancel vanilla es6 Promises right now, so you are forcing the user to use something like Bluebird to achieve this in the succesActionCreator.
My point is, maybe is difficult to implement but not add complexity for the user and make the api more powerful and attractive. I can help with the implementation of this if you decide to add the feature.

**Comment by ashishsc** [comment]
fwiw https://tc39.github.io/proposal-cancellation/

cancellation will probably eventually be native. It would be pretty annoying to require that for this though.

**Comment by ercpereda** [comment]
yes sure, when this be natively supported the `Cmd.cancel` could cancel the Promises as well, in the mean time, it could only ignore the responses of the Promises when they come.

**Comment by bdwain** [comment]
What i suggested does not require bluebird. It works with vanilla es6 promises. It also has the benefit of requiring few (if any) changes or concepts introduced to the api. I'd rather not introduce a large change to allow you to do something you can already do. 

My other worry is that trying to add a key to each Cmd makes it hard to isolate any given reducer.  In order to guarantee your key is unique (and avoid cancelling something you don't want to cancel), you have to know what all of the other keys in your app are. And it could get very confusing if you had a reducer that was used in multiple places at the same time.

**Comment by ercpereda** [comment]
I don't understand how you can accomplish that with vanilla es6 since there is not way to cancel a promise in es6. Maybe I missed out something in your suggestion.
About the `key` that is the same issue with the `type` of an action, you need to guarantee that the `type` is unique. Also that was only an example another approach can be something like:

```javascript
............
return loop(
 { type: "foo_action" },
 Cmd.run(foo, {
  successActionCreator,
  failActionCreator,
  cancelable: true
 }
);

Cmd.cancel("foo_action");
```

**Comment by bdwain** [comment]
My thing didn't cancel the promise. It just ignores the response if you update the state to ignore the promise. You can accomplish this by checking the state before resolving the promise and adding a property to the response or by checking the state in the reducer when handling the on success action. 

For example, if you had logic to update the status of an item, you could store the current updating id in the state. On update success, if the response id does not match the current id, ignore it and just return previous state.

**Comment by willisplummer** [comment]
I agree that cancelable async seems out of scope for `redux-loop` and resolvable on the users' end by simply tracking in state whether the reducer case for the `successAction` in question should return new state or just the old state.

e.g.

```
const reducer = (state) => {
  switch (action.type) {
    case 'HANDLE_FETCH_SUCCESS':
      state.cancelFetch ? state : { ...state, result: action.resultData }

      [ETC..]
  }
}
```

There are a number of ways a user could solve this without the use of any third party libraries.

**Comment by bdwain** [comment]
I'm going to close this issue because I don't think there's anything to do on it for now. I'd be happy to revisit the issue when Cancellations are natively supported.
